### ISSUE TITLE
Order confirmation email for guest checkout does not include download links

### ISSUE THREAD

**Issue by hsray** [magento/magento2#18323]
### Preconditions
1. Magento version: 2.2.5
2. PHP 7.0.x
3. Completely clean install with no demo data

### Steps to reproduce
1. Added one downloadable product with $0 price, shareable set to "Y" and unlimited downloads
2. Placed order as guest
3. Placed order as logged in customer

### Expected result
1. Email with download link for each order

### Actual result
1. Email for order placed when logged in was correct
2. Email for order placed as guest had two problems:
- the word "Links" was printed under the product SKU instead of "Download this" as set on the product
- no links were in the email
Screenprints of both emails are attached
![guest](https://user-images.githubusercontent.com/7255161/46259749-b1719080-c4b3-11e8-9eee-947cc1238337.JPG)
![logged-in-customer](https://user-images.githubusercontent.com/7255161/46259750-b1719080-c4b3-11e8-8e4d-40370df83509.JPG)

**Comment by magento-engcom-team** [comment]
Hi @hsray. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@hsray do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by hsray** [comment]
@magento-engcom-team give me 2.2.5 instance

**Comment by magento-engcom-team** [comment]
Hi @hsray. Thank you for your request. I'm working on Magento 2.2.5 instance for you

**Comment by magento-engcom-team** [comment]
Hi @hsray, here is your [Magento instance](https://i-18323-2-2-5.engcom.dev.magento.com).
Admin access: [https://i-18323-2-2-5.engcom.dev.magento.com/admin](https://i-18323-2-2-5.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by hsray** [comment]
I can confirm that I was able to reproduce the problem on the vanilla Magento instance. The logged in customer order email was correct and the guest order email was incorrect as described above.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-95909`, `MAGETWO-95910` were created

**Comment by mahesh-rajawat** [comment]
#SQUASHTOBERFEST

**Comment by magento-engcom-team** [comment]
Hi @maheshWebkul721. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by mahesh-rajawat** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @maheshWebkul721. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @maheshWebkul721, here is your [Magento instance](http://ec2-34-228-235-121.compute-1.amazonaws.com/i-18323-2-2-develop/).
Admin access: [http://ec2-34-228-235-121.compute-1.amazonaws.com/i-18323-2-2-develop//admin](http://ec2-34-228-235-121.compute-1.amazonaws.com/i-18323-2-2-develop//admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
Hi @swnsma. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by sidolov** [comment]
Hi @hsray. Thank you for your report. 
The issue has been fixed in magento/magento2#19040 by @swnsma in 2.3-develop branch
Related commit(s):
 - [5dc7add05b66271f4ad8c8be85c20d4beb173306](https://github.com/magento/magento2/commit/5dc7add05b66271f4ad8c8be85c20d4beb173306)
 - [e19d70e3fd43184eeeb009129bf0a564fae002ea](https://github.com/magento/magento2/commit/e19d70e3fd43184eeeb009129bf0a564fae002ea)

The fix will be available with the upcoming 2.3.1 release.

**Comment by magento-engcom-team** [comment]
Hi @hsray. Thank you for your report. 
The issue has been fixed in magento/magento2#19036 by @swnsma in 2.2-develop branch
Related commit(s):
 - [429815bee09f999ef3e38b93a5c097363fe2cbbd](https://github.com/magento/magento2/commit/429815bee09f999ef3e38b93a5c097363fe2cbbd)

The fix will be available with the upcoming 2.2.8 release.

**Comment by aholovan** [comment]
> Hi @magento-engcom-team.
> 
> * [429815bee09f999ef3e38b93a5c097363fe2cbbd](https://github.com/magento/magento2/commit/429815bee09f999ef3e38b93a5c097363fe2cbbd)
> 
> The fix will be available with the upcoming 2.2.8 release.

I've just checked current fix on Magento 2.3.0 as guest and registered customer and got from debug that fix of changing events doesn't work as needed. Namely SetLinkStatusObserver runs before than SaveDownloadableOrderItemObserver and that's why PurchasedLink has status 'pending', not 'available'. So, if you try download it as Guest you will get error that link not available.

**Comment by swnsma** [comment]
@aholovan,
Please create new issue for this case.

**Comment by djamps** [comment]
This change breaks download links for all new orders, not just guests.   Now, links are never set to available, cannot be downloaded whether the order is placed logged in or not.   We had to create a new module to override this change.

**Comment by swnsma** [comment]
@djamps 
You always can contribute your bugfix to Magento 2.
For now it look like that solution for your issue will be delivered in PR #21971

**Comment by djamps** [comment]
Thank you for the reply, for now our solution is to roll back this commit via custom module as none of our sites allow checkout as guest.   Hopefully PR [#21971](https://github.com/magento/magento2/pull/21971) fixes the guest problem without causing further issues.
